src/repositories/request-repository.ts

```diff
@@ -145,4 +150,68 @@ export class RequestRepository extends Repository<Request> {
+  /**
+   * Marks requests as READY if (in the order of precendence):
+   *  1. there are PROCESSING requests that need to be anchored and retried (the maximum anchoring delay has elapsed and the request hasn't been updated in a long time)
+   *  2. there are PENDING requests that need to be anchored (the maximum anchoring delay has elasped)
+   *  3. there are streamLimit streams needing an anchor (prioritizing PROCESSING requests that need to be retried, then PENDING requests)
+   * Returns the original requests that were marked as READY
+   */
+  public async findAndMarkReady(streamLimit: number): Promise<Request[]> {
+    const anchoringDeadline = new Date(Date.now() - MAX_ANCHORING_DELAY_MS)
+    const retryDeadline = new Date(Date.now() - MAX_ANCHORING_DELAY_MS / 2)
+    const isolationLevel =
+      this.connection.options.type === 'sqlite' ? 'SERIALIZABLE' : 'REPEATABLE READ'
+
+    return this.connection.transaction(isolationLevel, async (transactionalEntityManager) => {
+      // retrieves up to streamLimit unique streams with their earliest request createdAt value.
+      // this will only return streams associated with requests that are PENDING, or PROCESSING and needs to be retried
+      const streamsToAnchor = await transactionalEntityManager
+        .getRepository(Request)
+        .createQueryBuilder('request')
+        .select(['request.streamId', 'request.createdAt'])
+        .where('request.status = :processingStatus AND request.updatedAt < :retryDeadline', {
+          processingStatus: RequestStatus.PROCESSING,
+          retryDeadline: retryDeadline.toISOString(),
+        })
+        .orWhere('request.status = :pendingStatus', { pendingStatus: RequestStatus.PENDING })
+        .orderBy('MIN(request.createdAt)', 'ASC')
+        .groupBy('request.streamId')
+        .limit(streamLimit)
+        .getMany()
+
+      // Do not anchor if the earliest request isn't expired and there isn't enough streams
+      const earliestIsNotExpired =
+        streamsToAnchor.length > 0 && streamsToAnchor[0].createdAt > anchoringDeadline
+      if (earliestIsNotExpired && streamsToAnchor.length < streamLimit) {
+        return []
+      }
+
+      const streamIds = streamsToAnchor.map(({ streamId }) => streamId)
+
+      // retrieves all requests associated with the streams
+      const requests = await this.connection
+        .getRepository(Request)
+        .createQueryBuilder('request')
+        .orderBy('request.createdAt', 'ASC')
+        .where('request.streamId IN (:...streamIds)', { streamIds })
+        .getMany()
+
+      const results = await this.updateRequests(
+        { status: RequestStatus.READY },
+        requests,
+        transactionalEntityManager
+      )
+
+      // if not all requests are updated
+      if (!results.affected || results.affected != requests.length) {
+        throw Error(
+          `A problem occured when updated requests to READY. Only ${requests.length}/${results.affected} requests were updated`
```

Comment:
```suggestion
          `A problem occured when updated requests to READY. Only ${results.affected}/${requests.length} requests were updated`
```

Reply:
I think it's more common convention to show that only X out of an expected Y things happened as X/Y rather than Y/X

Reply:
haha you are totally right